8267190: Optimize Vector API test operations

sviswa7 opened the pull request:
Vector API test operations (IS_DEFAULT, IS_FINITE, IS_INFINITE, IS_NAN and IS_NEGATIVE) are computed in three steps:
1) reinterpreting the floating point vectors as integral vectors (int/long)
2) perform the test in integer domain to get a int/long mask
3) reinterpret the int/long mask as float/double mask
Step 3) currently is very slow. It can be optimized by modifying the Java code to utilize the existing reinterpret intrinsic.

For the VectorTestPerf attached to the JBS for JDK-8267190, the performance improves as follows:

Base:
Benchmark (size) Mode Cnt Score Error Units
VectorTestPerf.IS_DEFAULT 1024 thrpt 5 223.156 ± 90.452 ops/ms
VectorTestPerf.IS_FINITE 1024 thrpt 5 223.841 ± 91.685 ops/ms
VectorTestPerf.IS_INFINITE 1024 thrpt 5 224.561 ± 83.890 ops/ms
VectorTestPerf.IS_NAN 1024 thrpt 5 223.777 ± 70.629 ops/ms
VectorTestPerf.IS_NEGATIVE 1024 thrpt 5 218.392 ± 79.806 ops/ms

With patch:
Benchmark (size) Mode Cnt Score Error Units
VectorTestPerf.IS_DEFAULT 1024 thrpt 5 8812.357 ± 40.477 ops/ms
VectorTestPerf.IS_FINITE 1024 thrpt 5 7425.739 ± 296.622 ops/ms
VectorTestPerf.IS_INFINITE 1024 thrpt 5 8932.730 ± 269.988 ops/ms
VectorTestPerf.IS_NAN 1024 thrpt 5 8574.872 ± 498.649 ops/ms
VectorTestPerf.IS_NEGATIVE 1024 thrpt 5 8838.400 ± 11.849 ops/ms

Best Regards,
Sandhya

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267190](https://bugs.openjdk.java.net/browse/JDK-8267190): Optimize Vector API test operations


### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to b506fc45739f9a85586b3426b3cd4acf03b02383
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to bb0d400030bdefd98194af7d002d934dcba64d11


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4039/head:pull/4039` \
`$ git checkout pull/4039`

Update a local copy of the PR: \
`$ git checkout pull/4039` \
`$ git pull https://git.openjdk.java.net/jdk pull/4039/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4039`

View PR using the GUI difftool: \
`$ git pr show -t 4039`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4039.diff">https://git.openjdk.java.net/jdk/pull/4039.diff</a>

</details>

sviswa7 commented:
/label core-libs

PaulSandoz commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Byte128Vector.java:
```diff
@@ -599,10 +599,15 @@ Byte128Vector toVector() {
-        public <E> VectorMask<E> cast(VectorSpecies<E> s) {
-            AbstractSpecies<E> species = (AbstractSpecies<E>) s;
+        final <E>
+        VectorMask<E> defaultMaskReinterpret(VectorSpecies<E> dsp) {
```
Since this method is only called by `cast` we can make this method private and accept an argument of `AbstractSpecies<E>`. Further, the length check is duplicated by `cast` so we could turn it into an assert. Extra bonus points for converting the statement switch into an expression switch.

PaulSandoz reviewed (approved):
Changes look good, some minor comments. I shall run it through tier 1 to 3 tests and report back.

vnkozlov reviewed (approved):
Change in vectorIntrinsics.cpp seems fine. I did not look on Java code.

PaulSandoz commented:
Tier 1 to 3 tests pass on supported platforms

sviswa7 commented:
@PaulSandoz @vnkozlov Thanks a lot for the review. 
Paul, I have implemented your review comments. I also changed the switch to switch expression. Please take a look.

PaulSandoz commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-VectorBits.java.template:
```diff
@@ -842,31 +842,44 @@ final class $vectortype$ extends $abstractvectortype$ {
+        @ForceInline
+        private final <E>
+        VectorMask<E> defaultMaskCast(AbstractSpecies<E> dsp) {
+            boolean[] maskArray = toArray();
```
Can you add an `assert length() != species.laneCount()`?

PaulSandoz commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-VectorBits.java.template:
```diff
@@ -842,31 +842,44 @@ final class $vectortype$ extends $abstractvectortype$ {
+        VectorMask<E> defaultMaskCast(AbstractSpecies<E> dsp) {
+            boolean[] maskArray = toArray();
+            // enum-switches don't optimize properly JDK-8161245
+            return (
```
Minor syntactic quibble: you don't need the '(` and `)` surrounding the switch expressions e.g.:
```java
   return switch (dsp.laneType.switchKey) {
       case ...
   }
```

PaulSandoz reviewed (approved):
Java changes are good, some minor comments if you choose to accept them, no need for me to review further.

sviswa7 commented:
Thanks Paul. I have implemented these two suggestions as well.

If no objections from any one else, I plan to integrate this tomorrow, Friday May 21.

sviswa7 commented:
/integrate
